### DISCUSSION TITLE
Support for polymorphic send/publish?

### DISCUSSION THREAD

**Discussion by sreejith-ms** [BrighterCommand/Brighter#1612] [Q&A]
Hi,

My requirements are very similar to https://github.com/BrighterCommand/Brighter/issues/116 and https://github.com/BrighterCommand/Brighter/issues/144.

I'm trying to switch from MediatR to Brighter. I have a wrapper class (`CommandHandlerBase`), used for the command handler delegation to MediatR. The command handlers in the application use the wrapper class (so it is easy to switch to another library). Also, there are base classes used for commands and domain events.

The commands and events are using this base class/interface to send/publish. Brighter throws an exception as `No command handler was found for the typeof command paramore.brighter.commandprocessor.ICommand - a command should have exactly one handler`

Examples:

```
var domainEvents = domainEventsAccessor.GetAllDomainEvents();

foreach (IDomainEvent domainEvent in domainEvents)
{
    await _commandProcessor.Publish(domainEvent);
}           
```

```
ICommand command = new GreetingCommand("Ian");
await _commandProcessor.SendAsync(command, cancellationToken: cancellationToken);
```

`Found 0 async pipelines for event: HelloAsyncListeners.GreetingEvent`

The domain events are stored in the entity, similar to: https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/Services/Ordering/Ordering.Domain/SeedWork/Entity.cs#L23
and the events are published before committing https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/Services/Ordering/Ordering.Infrastructure/OrderingContext.cs#L63

I think Brighter uses generics to resolve the type from the registry, but Mediator uses the type to resolve.
https://github.com/BrighterCommand/Brighter/blob/master/src/Paramore.Brighter/Interpreter.cs#L61
https://github.com/jbogard/MediatR/blob/master/src/MediatR/Mediator.cs#L36

Any thoughts/workaround on this? apart from the nasty `(dynamic)command` casting.

**Comment by iancooper** [comment]
So if you do:

await _commandProcessor.Publish((dynamic)domainEvent);

This should work as expected. That is what we use most places.

I note that you say you are against that. I guess your concern is that it's not a 'pit of success' for folks, who have to know this workaround on generic collections.

We don't have an alternative now, although we could look at this question again if it is enough of a problem.

**Reply by sreejith-ms** [reply]
Thanks @iancooper

**Reply by iancooper** [reply]
@sreejith-ms If you are looking at using Brighter for eShopOnContainers, I'd wait until v9 comes out of pre-release. We will have better Outbox support, and integration with ASP.NET Scopes is likely to work better.
